### PR TITLE
Fix the Rank of logsumexp Tensor and mGPU support.

### LINKED ISSUE #131316: SDPA + compile + bfloat16 fails (ROCm)
### 🐛 Describe the bug

Using SDPA with torch.compile in bfloat16 causes an error, perhaps only with ROCm/HIP. I don't have Nvidia GPUs for testing. Not compiling the model, or using float32 instead of bfloat16, makes the code function.

```python
import torch
import torch.nn.functional as F
from torch import nn

class ModelArgs:
    dim: int = 512
    n_heads: int = 8
    n_kv_heads: int = 8
    max_seq_len: int = 128

class Attention(nn.Module):
    def __init__(self, args: ModelArgs):
        super().__init__()
        self.n_heads = args.n_heads
        self.n_kv_heads = args.n_kv_heads
        self.head_dim = args.dim // args.n_heads
        
        self.wq = nn.Linear(args.dim, args.n_heads * self.head_dim, bias=False)
        self.wk = nn.Linear(args.dim, args.n_kv_heads * self.head_dim, bias=False)
        self.wv = nn.Linear(args.dim, args.n_kv_heads * self.head_dim, bias=False)
        self.wo = nn.Linear(args.n_heads * self.head_dim, args.dim, bias=False)

    def forward(self, x: torch.Tensor):
        bs, seqlen, _ = x.shape
        xq, xk, xv = self.wq(x), self.wk(x), self.wv(x)
        
        xq = xq.view(bs, seqlen, self.n_heads, self.head_dim)
        xk = xk.view(bs, seqlen, self.n_kv_heads, self.head_dim)
        xv = xv.view(bs, seqlen, self.n_kv_heads, self.head_dim)

        xq = xq.transpose(1, 2)
        xk = xk.transpose(1, 2)
        xv = xv.transpose(1, 2)

        output = F.scaled_dot_product_attention(xq, xk, xv, is_causal=True)
        output = output.transpose(1, 2).contiguous().view(bs, seqlen, -1)
        return self.wo(output)

def main():
    torch.manual_seed(0)
    device = torch.device("cuda")
    torch.set_default_dtype(torch.bfloat16)

    args = ModelArgs()
    model = Attention(args).to(device)
    model = torch.compile(model)

    x = torch.randn(1, args.max_seq_len, args.dim, requires_grad=True, device=device)
    x = x.bfloat16()

    output = model(x)
    loss = output.sum()
    loss.backward()

    print("Input gradient shape:", x.grad.shape)
    print("Input gradient sum:", x.grad.sum().item())
    print("Output shape:", output.shape)
    print("Output sum:", output.sum().item())

if __name__ == "__main__":
    main()
```

The traceback:
```Traceback (most recent call last):
  File "/scratch/project_465000670/maester/minimum_repro.py", line 69, in <module>
    main()
  File "/scratch/project_465000670/maester/minimum_repro.py", line 61, in main
    loss.backward()
  File "/scratch/project_465000670/maester/.venv/lib/python3.10/site-packages/torch/_tensor.py", line 522, in backward
    torch.autograd.backward(
  File "/scratch/project_465000670/maester/.venv/lib/python3.10/site-packages/torch/autograd/__init__.py", line 346, in backward
    _engine_run_backward(
  File "/scratch/project_465000670/maester/.venv/lib/python3.10/site-packages/torch/autograd/graph.py", line 806, in _engine_run_backward
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/scratch/project_465000670/maester/.venv/lib/python3.10/site-packages/torch/autograd/function.py", line 306, in apply
    return user_fn(self, *args)
  File "/scratch/project_465000670/maester/.venv/lib/python3.10/site-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 1976, in backward
    out = call_compiled_backward()
  File "/scratch/project_465000670/maester/.venv/lib/python3.10/site-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 1915, in call_compiled_backward
    out = call_func_at_runtime_with_args(
  File "/scratch/project_465000670/maester/.venv/lib/python3.10/site-packages/torch/_functorch/_aot_autograd/utils.py", line 120, in call_func_at_runtime_with_args
    out = normalize_as_list(f(args))
  File "/scratch/project_465000670/maester/.venv/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 602, in _fn
    return fn(*args, **kwargs)
  File "/scratch/project_465000670/maester/.venv/lib/python3.10/site-packages/torch/_inductor/codecache.py", line 1212, in __call__
    return self.current_callable(inputs)
  File "/scratch/project_465000670/maester/.venv/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 991, in run
    return model(new_inputs)
  File "/tmp/torchinductor_larsenra/63/c63nhtwlwgadmoet2eag5f5hk6etntppfn6klu2r6ma324dybsvj.py", line 81, in call
    assert_size_stride(getitem_1, (1, 8, 128), (1024, 128, 1))
AssertionError: wrong number of dimensions
```

The generated code:
```python

# AOT ID: ['0_backward']
from ctypes import c_void_p, c_long
import torch
import math
import random
import os
import tempfile
from math import inf, nan
from torch._inductor.hooks import run_intermediate_hooks
from torch._inductor.utils import maybe_profile
from torch._inductor.codegen.memory_planning import _align as align

from torch import device, empty_strided
from torch._inductor.async_compile import AsyncCompile
from torch._inductor.select_algorithm import extern_kernels
from torch._inductor.codegen.multi_kernel import MultiKernelCall

aten = torch.ops.aten
inductor_ops = torch.ops.inductor
_quantized = torch.ops._quantized
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
empty_strided_cpu = torch._C._dynamo.guards._empty_strided_cpu
empty_strided_cuda = torch._C._dynamo.guards._empty_strided_cuda
reinterpret_tensor = torch._C._dynamo.guards._reinterpret_tensor
alloc_from_pool = torch.ops.inductor._alloc_from_pool
async_compile = AsyncCompile()


# kernel path: /tmp/torchinductor_larsenra/hk/chkkhpdnd4hzy5dv7x545hlv6sypde24g6vkdrk6sesiu6ce5l4q.py
# Source Nodes: [], Original ATen: [aten.add]

triton_poi_fused_add_0 = async_compile.triton('triton_', '''
import triton
import triton.language as tl
from triton.compiler.compiler import AttrsDescriptor

from torch._inductor.runtime import triton_helpers, triton_heuristics
from torch._inductor.runtime.triton_helpers import libdevice, math as tl_math
from torch._inductor.runtime.hints import AutotuneHint, ReductionHint, TileHint, instance_descriptor, DeviceProperties

@triton_heuristics.pointwise(
    size_hints=[65536], 
    filename=__file__,
    triton_meta={'signature': {0: '*bf16', 1: '*bf16', 2: '*bf16', 3: 'i32'}, 'device': DeviceProperties(type='hip', index=0, cc='gfx90a', major=None, regs_per_multiprocessor=None, max_threads_per_multi_processor=None, multi_processor_count=None), 'constants': {}, 'configs': [AttrsDescriptor(divisible_by_16=(0, 1, 2, 3), equal_to_1=())]},
    inductor_meta={'autotune_hints': set(), 'kernel_name': 'triton_poi_fused_add_0', 'mutated_arg_names': ['in_out_ptr0'], 'no_x_dim': False, 'num_load': 3, 'num_reduction': 0, 'backend_hash': None, 'are_deterministic_algorithms_enabled': False, 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': False, 'force_disable_caches': False, 'dynamic_scale_rblock': True, 'max_autotune': False, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False, 'is_hip': True},
    min_elem_per_thread=0
)
@triton.jit
def triton_(in_out_ptr0, in_ptr0, in_ptr1, xnumel, XBLOCK : tl.constexpr):
    xnumel = 65536
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = tl.full([XBLOCK], True, tl.int1)
    x0 = xindex
    tmp0 = tl.load(in_ptr0 + (x0), None).to(tl.float32)
    tmp1 = tl.load(in_ptr1 + (x0), None).to(tl.float32)
    tmp3 = tl.load(in_out_ptr0 + (x0), None).to(tl.float32)
    tmp2 = tmp0 + tmp1
    tmp4 = tmp2 + tmp3
    tl.store(in_out_ptr0 + (x0), tmp4, None)
''', device_str='cuda')

import triton
import triton.language as tl
from torch._inductor.runtime.triton_heuristics import grid, split_scan_grid, start_graph, end_graph
from torch._C import _cuda_getCurrentRawStream as get_raw_stream


async_compile.wait(globals())
del async_compile

def call(args):
    view, permute_3, permute_4, permute_5, getitem, getitem_1, getitem_6, getitem_7, permute_10, permute_18, permute_22, permute_26, tangents_1 = args
    args.clear()
    assert_size_stride(view, (128, 512), (512, 1))
    assert_size_stride(permute_3, (1, 8, 128, 64), (65536, 64, 512, 1))
    assert_size_stride(permute_4, (1, 8, 128, 64), (65536, 64, 512, 1))
    assert_size_stride(permute_5, (1, 8, 128, 64), (65536, 64, 512, 1))
    assert_size_stride(getitem, (1, 8, 128, 64), (65536, 64, 512, 1))
    assert_size_stride(getitem_1, (1, 8, 128), (1024, 128, 1))
    assert_size_stride(getitem_6, (), ())
    assert_size_stride(getitem_7, (), ())
    assert_size_stride(permute_10, (512, 512), (512, 1))
    assert_size_stride(permute_18, (512, 512), (512, 1))
    assert_size_stride(permute_22, (512, 512), (512, 1))
    assert_size_stride(permute_26, (512, 512), (512, 1))
    assert_size_stride(tangents_1, (1, 128, 512), (65536, 512, 1))
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0)
        buf0 = empty_strided_cuda((512, 512), (512, 1), torch.bfloat16)
        # Source Nodes: [], Original ATen: [aten.mm]
        extern_kernels.mm(reinterpret_tensor(tangents_1, (512, 128), (1, 512), 0), reinterpret_tensor(getitem, (128, 512), (512, 1), 0), out=buf0)
        buf1 = empty_strided_cuda((128, 512), (512, 1), torch.bfloat16)
        # Source Nodes: [], Original ATen: [aten.mm]
        extern_kernels.mm(reinterpret_tensor(tangents_1, (128, 512), (512, 1), 0), permute_10, out=buf1)
        del permute_10
        del tangents_1
        # Source Nodes: [], Original ATen: [aten._scaled_dot_product_flash_attention_backward]
        buf2 = aten._scaled_dot_product_flash_attention_backward.default(reinterpret_tensor(buf1, (1, 8, 128, 64), (65536, 64, 512, 1), 0), permute_3, permute_4, permute_5, getitem, getitem_1, None, None, 128, 128, 0.0, True, getitem_6, getitem_7, scale=0.125)
        del getitem
        del getitem_1
        del getitem_6
        del getitem_7
        del permute_3
        del permute_4
        del permute_5
        buf3 = buf2[0]
        buf4 = buf2[1]
        buf5 = buf2[2]
        del buf2
        buf6 = empty_strided_cuda((512, 512), (512, 1), torch.bfloat16)
        # Source Nodes: [], Original ATen: [aten.mm]
        extern_kernels.mm(reinterpret_tensor(buf5, (512, 128), (1, 512), 0), view, out=buf6)
        buf7 = buf1; del buf1  # reuse
        # Source Nodes: [], Original ATen: [aten.mm]
        extern_kernels.mm(reinterpret_tensor(buf5, (128, 512), (512, 1), 0), permute_18, out=buf7)
        del permute_18
        buf8 = empty_strided_cuda((512, 512), (512, 1), torch.bfloat16)
        # Source Nodes: [], Original ATen: [aten.mm]
        extern_kernels.mm(reinterpret_tensor(buf4, (512, 128), (1, 512), 0), view, out=buf8)
        buf9 = reinterpret_tensor(buf5, (128, 512), (512, 1), 0); del buf5  # reuse
        # Source Nodes: [], Original ATen: [aten.mm]
        extern_kernels.mm(reinterpret_tensor(buf4, (128, 512), (512, 1), 0), permute_22, out=buf9)
        del permute_22
        buf10 = empty_strided_cuda((512, 512), (512, 1), torch.bfloat16)
        # Source Nodes: [], Original ATen: [aten.mm]
        extern_kernels.mm(reinterpret_tensor(buf3, (512, 128), (1, 512), 0), view, out=buf10)
        del view
        buf11 = reinterpret_tensor(buf4, (128, 512), (512, 1), 0); del buf4  # reuse
        # Source Nodes: [], Original ATen: [aten.mm]
        extern_kernels.mm(reinterpret_tensor(buf3, (128, 512), (512, 1), 0), permute_26, out=buf11)
        del buf3
        del permute_26
        buf12 = reinterpret_tensor(buf11, (1, 128, 512), (65536, 512, 1), 0); del buf11  # reuse
        # Source Nodes: [], Original ATen: [aten.add]
        stream0 = get_raw_stream(0)
        triton_poi_fused_add_0.run(buf12, buf7, buf9, 65536, grid=grid(65536), stream=stream0)
        del buf7
        del buf9
    return (reinterpret_tensor(buf10, (512, 512), (512, 1), 0), reinterpret_tensor(buf8, (512, 512), (512, 1), 0), reinterpret_tensor(buf6, (512, 512), (512, 1), 0), reinterpret_tensor(buf0, (512, 512), (512, 1), 0), buf12, )


def benchmark_compiled_module(times=10, repeat=10):
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    view = rand_strided((128, 512), (512, 1), device='cuda:0', dtype=torch.bfloat16)
    permute_3 = rand_strided((1, 8, 128, 64), (65536, 64, 512, 1), device='cuda:0', dtype=torch.bfloat16)
    permute_4 = rand_strided((1, 8, 128, 64), (65536, 64, 512, 1), device='cuda:0', dtype=torch.bfloat16)
    permute_5 = rand_strided((1, 8, 128, 64), (65536, 64, 512, 1), device='cuda:0', dtype=torch.bfloat16)
    getitem = rand_strided((1, 8, 128, 64), (65536, 64, 512, 1), device='cuda:0', dtype=torch.bfloat16)
    getitem_1 = rand_strided((1, 8, 128), (1024, 128, 1), device='cuda:0', dtype=torch.float32)
    getitem_6 = rand_strided((), (), device='cuda:0', dtype=torch.int64)
    getitem_7 = rand_strided((), (), device='cuda:0', dtype=torch.int64)
    permute_10 = rand_strided((512, 512), (512, 1), device='cuda:0', dtype=torch.bfloat16)
    permute_18 = rand_strided((512, 512), (512, 1), device='cuda:0', dtype=torch.bfloat16)
    permute_22 = rand_strided((512, 512), (512, 1), device='cuda:0', dtype=torch.bfloat16)
    permute_26 = rand_strided((512, 512), (512, 1), device='cuda:0', dtype=torch.bfloat16)
    tangents_1 = rand_strided((1, 128, 512), (65536, 512, 1), device='cuda:0', dtype=torch.bfloat16)
    fn = lambda: call([view, permute_3, permute_4, permute_5, getitem, getitem_1, getitem_6, getitem_7, permute_10, permute_18, permute_22, permute_26, tangents_1])
    return print_performance(fn, times=times, repeat=repeat)


if __name__ == "__main__":
    from torch._inductor.wrapper_benchmark import compiled_module_main
    compiled_module_main('None', benchmark_compiled_module)
```

### Versions

PyTorch version: 2.5.0.dev20240720+rocm6.0
Is debug build: False
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: 6.0.32830-d62f6a171

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 17.0.0 (https://github.com/RadeonOpenCompute/llvm-project roc-6.0.2 24012 af27734ed982b52a9f1be0f035ac91726fc697e4)
CMake version: version 3.26.4
Libc version: glibc-2.35

Python version: 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.14.21-150400.24.81_12.0.75-cray_shasta_c-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: AMD Instinct MI250X (gfx90a:sramecc+:xnack-)
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: 6.0.32830
MIOpen runtime version: 3.0.0
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             128
On-line CPU(s) list:                0-127
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 7A53 64-Core Processor
CPU family:                         25
Model:                              48
Thread(s) per core:                 2
Core(s) per socket:                 64
Socket(s):                          1
Stepping:                           1
Frequency boost:                    enabled
CPU max MHz:                        2000.0000
CPU min MHz:                        1500.0000
BogoMIPS:                           3992.47
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm
Virtualization:                     AMD-V
L1d cache:                          2 MiB (64 instances)
L1i cache:                          2 MiB (64 instances)
L2 cache:                           32 MiB (64 instances)
L3 cache:                           256 MiB (8 instances)
NUMA node(s):                       4
NUMA node0 CPU(s):                  0-15,64-79
NUMA node1 CPU(s):                  16-31,80-95
NUMA node2 CPU(s):                  32-47,96-111
NUMA node3 CPU(s):                  48-63,112-127
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Mitigation; safe RET, no microcode
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] pytorch-triton-rocm==3.0.0+21eae954ef
[pip3] torch==2.5.0.dev20240720+rocm6.0
[pip3] torchdata==0.7.1.dev20240707+cpu
[pip3] torchvision==0.1.6
[conda] No relevant packages

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @ezyang @anijain2305 @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by xinyazhang**
The logsumexp tensor was considered for internal use only but apparently exposed to unit tests and inductors.

The stream should be selected after picking the current device. Otherwise the code is checking the default device's architecture.

Fixes #131316 #137414

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -16266,14 +16266,14 @@ def sample_inputs_alias_copy(op_info, device, dtype, requires_grad, **kwargs):
                          device_type='cuda', dtypes=(torch.bfloat16, torch.float16, torch.float32),
                          active_if=TEST_WITH_ROCM and PLATFORM_SUPPORTS_MEM_EFF_ATTENTION),
             # for element 1, was torch.Size([4, 4, 0]) but real shape was torch.Size([16, 3, 0])
-            DecorateInfo(unittest.expectedFailure, "TestMeta", "test_dispatch_meta_outplace", device_type="cuda",
+            DecorateInfo(unittest.skip, "TestMeta", "test_dispatch_meta_outplace", device_type="cuda",
```
Comment (malfet) [inline torch/testing/_internal/common_methods_invocations.py:16269]: Why are you replacing expectedFailure with skip here?

**Review by malfet — CHANGES_REQUESTED** [review]
Please explain why one wants to skip tests now rather than `xfail`? Otherwise looks good to me.
Though, why are you pinging me rather than  @drisspg review on flash attention changes?

**Comment by jataylo** [comment]
> Please explain why one wants to skip tests now rather than `xfail`? Otherwise looks good to me. Though, why are you pinging me rather than @drisspg review on flash attention changes?

Thanks, added @drisspg as a reviewer now.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (jataylo) [reply torch/testing/_internal/common_methods_invocations.py]: We didn't understand why they were listed XFAIL in the first place, seems that we should list as skipped instead to move towards enabling these at some point. We can maintain XFAIL though if that is preferred here.

Comment (jataylo) [reply torch/testing/_internal/common_methods_invocations.py]: Moved back to XFAIL for now.

**Review by drisspg — APPROVED** [review]

**Pull request closed without merging**